polaris/utils/progress_indicator.py

```diff
@@ -0,0 +1,32 @@
+from contextlib import contextmanager
+from yaspin import yaspin
+from yaspin.spinners import Spinners
+
+
+@contextmanager
+def progress_indicator(start_msg: str = "In progress...", success_msg: str = "", error_msg: str = ""):
+    spinner = yaspin()
+    spinner.spinner = Spinners.dots
+    spinner.text = start_msg
+    result_message = {"message": ""}
+
+    try:
+        spinner.start()
+        yield result_message
```

Comment:
Any thoughts on using this to pass a message back up to the context manager? This preserves the current flow we have for upload operations but might seem a little hacky(?) 

Let me know what you think 😄

Reply:
@Andrewq11 How about yielding a callback?

```python
@contextmanager
def progress_indicator(start_msg: str = 'In progress...', success_msg: str | None = None, error_msg: str | None = None):
    spinner = yaspin(Spinners.dots, text=start_msg)

    result_message = {"message": success_msg}

    def update_message(msg):
        result_message["message"] = msg

    try:
        spinner.start()
        yield update_message
        spinner.text = ""
        spinner.green.ok(f"✅ SUCCESS: \033[1m{result_message['message']}\033[0m\n")
    except Exception as err:
        spinner.text = "" 
        spinner.red.fail(f"💥 ERROR: {error_msg}")
        raise err
    finally:
        spinner.stop()


with progress_indicator("Starting task...", success_msg="Task completed!", error_msg="Task failed!") as update_message:
    # Simulate some work
    import time
    time.sleep(2)
    # raise Exception

    update_message("All steps were successfully executed.")
```

This keeps the logic inside the `progress_indicator` function, and you can use the `update_message` callback to set a custom success message.

Reply:
I like it! Will update.

Reply:
Returning a callback function from a context manager feels a little un-Pythonic to me. 

To offer an alternative: You could also implement the context manager as a class. In this case, you need to implement the `__enter()__` and `__exit__()` magic methods. The enter method could then return `self` and you can add the method to the class.

Reply:
```py

class ProgressIndicator:
    def __init__(self, start_msg: str = 'In progress...', success_msg: str | None = None, error_msg: str | None = None):
        self._start_msg = start_msg
        self._success_msg = success_msg
        self._error_msg = error_msg

    def __enter__(self):
        spinner = yaspin(Spinners.dots, text=self._start_msg)
        spinner.start()
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        spinner.text = "" 
        if exc_type is not None: 
            spinner.red.fail(f"💥 ERROR: {self._error_msg}")
        else: 
             spinner.green.ok(f"✅ SUCCESS: \033[1m{self._success_msg}\033[0m\n")
        spinner.stop()

    def update_success_msg(self, msg: str): 
        self._success_msg = msg


with ProgressIndicator("Starting task...", success_msg="Task completed!", error_msg="Task failed!") as indicator:
    indicator.update_success_msg("All steps were successfully executed.")
```

Reply:
I personally prefer the original in terms of readability, but that is likely because I'm not used to the pattern you suggested. I do agree that yielding a function doesn't seem very Pythonic though. If we feel strongly about this, I'm happy to implement it.

Reply:
My preference is the latter. Also would make it easy to add additional functionality to the context, such as `tqdm` integration or setting a custom error message. 

No strong preference though. We can add that complexity once it becomes needed as well.

Reply:
Let's do that then, not the biggest lift to implement it now.